File: bin/water.cpp
/**
 * @file	urban.cpp
 * @author	Jesús Carabaño Bravo <jcaraban@abo.fi>
 *
 * Cellular automata for water flow
 */

#include "../map.hpp"
using namespace map;


int main(int argc, char **argv)
{	
	//// Arguments

	assert(argc > 5);

	//// Configuration

	setupDevices("",DEV_GPU,"");
	//setupDevices("Intel",DEV_CPU,"");

	//// Inputs

	auto h = read(argv[1]); // digital elevation layer
	auto w = read(argv[2]); // water level layer
	auto i = read(argv[3]); // inlets inflow layer
	auto o = read(argv[4]); // outlets outflow layer
	Raster l;

	auto N = (argc > 6) ? atoi(argv[6]) : 1 ;

	auto swap = [&](std::vector<Raster> &x, int i, int j) {
		auto aux = max(x[i],x[j]);
		x[i] = min(x[i],x[j]);
		x[j] = aux;
	};

	auto netsort5 = [&](std::vector<Raster> &x) {
		swap(x,0,1); swap(x,2,3); swap(x,0,2);
		swap(x,3,4); swap(x,0,3); swap(x,1,3);
		swap(x,2,4); swap(x,1,4); swap(x,1,2);
	};

	auto avglevel = [&](Raster w, Raster h, std::vector<Raster> &x) {
		netsort5(x);     // in ascending order
		auto e = 0.001f; // epsilon
		auto s = w+x[0]; // sum variable
		auto c = ones(); // count variable
		for (int i=1; i<5; i++) {
			auto b = (s-e >= x[i]*i);
			s = s + b*x[i];
			c = c + b;
		}
		return s / c;
	};

	auto gather = [&](Raster w, Raster h) {
		std::vector<Raster> x(5); // list of neighbors
		x[0] = h; // central cell height
		x[1] = (h+w)({0,-1});
		x[2] = (h+w)({-1,0});
		x[3] = (h+w)({+1,0});
		x[4] = (h+w)({0,+1});
		return avglevel(w,h,x);
	};

	auto distrib = [&](Raster w, Raster h, Raster l) {
		auto wh = w+h;    // prev water level
		auto c = zeros(); // change in water
		c = c + max(0.0f, l({0,-1}) - wh);
		c = c + max(0.0f, l({-1,0}) - wh);
		c = c + max(0.0f, l({+1,0}) - wh);
		c = c + max(0.0f, l({0,+1}) - wh);
		c = c + max(h, l) - wh;
		auto cwh = max(c + wh, h);
		return cwh - h;
	};

	auto inflow = [&](Raster w, Raster i) {
		return w + i;
	};

	auto outflow = [&](Raster w, Raster o) {
		return max(w - o, zeros());
	};

	// Main Loop

	for (int j=0; j<N; j++) {
		w = inflow(w,i);    // add water to the inlets
		l = gather(w,h);    // gather the avg. water level
		w = distrib(w,h,l); // distribute the water
		w = outflow(w,o);   // remove water from the outlets
	}

	write(w,argv[5]);
}

File: python/water.py
from map import * ## "Parallel Map Algebra" package

w = read('water')   # water depth layer
h = read('dem') - w # digital elevation layer
i = read('inflow')  # inlets inflow layer
o = read('outflow') # outlets outflow layer
T = 1000            # number of time steps

def swap(x,i,j):
	x[i], x[j] = min(x[i],x[j]), max(x[i],x[j])

def netsort5(x):
	swap(x,0,1); swap(x,2,3); swap(x,0,2)
	swap(x,3,4); swap(x,0,3); swap(x,1,3)
	swap(x,2,4); swap(x,1,4); swap(x,1,2)

def avglevel(w,h,x):                 #   Minimization   #
	netsort5(x) # ascending order    #  of differences, #
	s = w+x[0]  # sum variable       # Di Gregorio 1999 #
	n = 1       # count variable     #                  #
	for i in range(1,5):             # Central cell:    #
		b = (s >= x[i]*i)            # 7 h + 11 w       #
		s += b*x[i]                  #                  #
		n += b                       # Von Neumann NBH: #
	return s / n                     #      | 30 |      #
                                     # | 13 |  7 |  6 | #
def gather(w,h):                     #      |  3 |      #
	x = [0]*5 # neigborhood (NBH)    #                  #
	x[0] = h  # central cell         # Sorting:         #
	x[1] = h(0,-1) + w(0,-1)         # |3|6|7|13|30|    #
	x[2] = h(-1,0) + w(-1,0)         #                  #
	x[3] = h(+1,0) + w(+1,0)         # Exclusion:       #
	x[4] = h(0,+1) + w(0,+1)         # |3|6|7|*|*|      #
	return avglevel(w,h,x)           #                  #
                                     # Gathered avg.    #
def distri(w,h,l):                   # water level: 9   #
	wh = w+h # prev water level      #                  #
	c  = max(0, l(0,-1) − wh)        # Water distrib:   #
	c += max(0, l(-1,0) − wh)        # |3+6|6+3|7+2|*|* #
	c += max(0, l(+1,0) − wh)        #                  #
	c += max(0, l(0,+1) - wh)        # Final chw level: #
	c += max(h, l) − wh              #      | 30 |      #
	cwh = max(c + wh, h)             # | 13 |  9 |  9 | #
	return cwh - h                   #      |  9 |      #

for t in range(T):
	w = w + i         # fill inlets with water
	l = gather(w,h)   # gather avg. water level
	w = distri(w,h,l) # distribute water to NBH
	w = max(w-o,0)    # drain water from outlets

write(w,'output')
